artichoke-backend/src/encoding_registry.rs

```diff
@@ -0,0 +1,160 @@
+use core::ffi::c_void;
+use regex::Regex;
+use spinoso_exception::TypeError;
+use std::sync::OnceLock;
+
+use crate::core::EncodingRegistry;
+use crate::def::NotDefinedError;
+use crate::encoding;
+use crate::error::Error;
+use crate::ffi::InterpreterExtractError;
+use crate::sys;
+use crate::types::Ruby;
+use crate::value::Value;
+use crate::Artichoke;
+use artichoke_core::constant::DefineConstant;
+use artichoke_core::value::Value as _;
+use std::ptr;
+
+impl EncodingRegistry for Artichoke {
+    type Value = Value;
+    type Error = Error;
+    type Spec = encoding::Spec;
+
+    fn def_encoding(&mut self, spec: Self::Spec) -> Result<(), Error> {
+        // Setup the class name regex
+        static ESCAPABLE_CONSTANT_CHARS: OnceLock<Regex> = OnceLock::new();
+        let escapable_const_char_regex = ESCAPABLE_CONSTANT_CHARS.get_or_init(|| {
+            // Match any `-` and `.` chars. These will be replaced with `_`
+            Regex::new(r"[-\.]+").unwrap()
```

Comment:
regex is really heavy here. We should be able to use `ByteSlice::replace` from `bstr`: https://docs.rs/bstr/latest/bstr/trait.ByteSlice.html#method.replace

Reply:
Ah nice, that is indeed much better.

I was thinking of whether there is another way to create bare symbols. e.g.

```ruby
irb(main):001:0> constants = Encoding.constants
=> 
[:ISO8859_14,                                                         
...                                                                   
irb(main):002:0> Encoding::UTF_8.replicate("Super Special Encoding")
=> #<Encoding:Super Special Encoding>
irb(main):003:0> Encoding.constants - constants
=> [:Super_Special_Encoding, :SUPER_SPECIAL_ENCODING]
irb(main):004:0> Encoding::Super_Special_Encoding
=> #<Encoding:Super Special Encoding>
irb(main):005:0> Encoding::Super_Special_Encoding.name
=> "Super Special Encoding"
```

And check this out:

```ruby
irb(main):007:0> Encoding::UTF_8.replicate("\x7F")
=> #<Encoding:>
irb(main):008:0> Encoding::UTF_8.replicate("\x80")
(Object doesn't support #inspect)
=>                                                                                
irb(main):009:0> _.name
=> "\x80"
irb(main):010:0> _.encoding
=> #<Encoding:US-ASCII>
irb(main):011:0> Encoding.find("\x80")
(Object doesn't support #inspect)
=>                                                               
irb(main):012:0> _.name
=> "\x80"
```

Valid names apparently need to be asciicompat as per https://github.com/artichoke/artichoke/blob/trunk/artichoke-backend/vendor/ruby/include/ruby/internal/encoding/encoding.h#L766-L793 which is used here: https://github.com/artichoke/artichoke/blob/trunk/artichoke-backend/vendor/ruby/encoding.c#L288-L301

Eventually some more to look into maybe, but:

```ruby
irb(main):001:0> constants = Encoding.constants
=> 
[:ISO8859_14,                                                          
...                                                                    
irb(main):002:0> (1..127).map(&:chr).each_slice(32).map { |c| Encoding::UTF_8.replicate(c.join) }
=> 
[#<Encoding:                                                           
                                                                       
                                                                       
 >,                                                                    
 #<Encoding:!"#$%&'()*+,-./0123456789:;<=>?@>,                         
 #<Encoding:ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`>,                         
 #<Encoding:abcdefghijklmnopqrstuvwxyz{|}~>]                           
irb(main):003:0> Encoding.constants - constants
=> [:ABCDEFGHIJKLMNOPQRSTUVWXYZ______, :ABCDEFGHIJKLMNOPQRSTUVWXYZ_____]
```

(Had to chunk it because otherwise some issue with Constants and lengths that I think it was getting mad at).

Reply:
Note: in the between time, going to take the easy Bstr replace approach.